### DISCUSSION TITLE
Chaining target factories: target not found

### DISCUSSION THREAD

**Discussion by prototaxites** [ropensci/targets#1162] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Trying to build a set of target factories to abstract away some repetitive tasks I'll need to run a lot. I am quite new to the concept of metaprogramming, so there's probably something obvious I've missed. 

There are two factories: the first (tar_read_ampliseq) reads in the data and creates an output (phyloseq) object, and the second (tar_decontaminate_phyloseq) does some processing of this object downstream. However, the second factory gives me an error saying the object is not found. Running the first target factory on its own works, and the target is listed in `tar_manifest()`. I can also access the target subsequently with a call to `tar_target()` - just not my second factory:

Target factories:
```r
tar_read_ampliseq <- function(name, table, taxonomy, metadata) {
  tar_name <- deparse(substitute(name))
  name_table_file <- paste0(tar_name, "_table_file")
  name_taxonomy_file <- paste0(tar_name, "_taxonomy_file")
  name_metadata_file <- paste0(tar_name, "_metadata_file")
  name_table <- paste0(tar_name, "_table")
  name_taxonomy <- paste0(tar_name, "_taxonomy")
  name_metadata <- paste0(tar_name, "_metadata")
  name_phyloseq <- paste0(tar_name, "_phyloseq")
  list(
    tar_target_raw(name_table_file, table, format = "file"),
    tar_target_raw(name_taxonomy_file, taxonomy, format = "file"),
    tar_target_raw(name_metadata_file, metadata, format = "file"),
    tar_target_raw(name_table, 
                   substitute(read_tsv(file, col_types = cols()),
                   env = list(file = as.symbol(name_table_file)))),
    tar_target_raw(name_taxonomy, 
                   substitute(read_tsv(file, col_types = cols()),
                              env = list(file = as.symbol(name_taxonomy_file)))),
    tar_target_raw(name_metadata, 
                   substitute(read_tsv(file, col_types = cols()),
                              env = list(file = as.symbol(name_metadata_file)))),
    tar_target_raw(name_phyloseq, substitute(
      build_phyloseq(table, taxonomy, metadata),
      env = list(table = as.symbol(name_table),
                 taxonomy = as.symbol(name_taxonomy),
                 metadata = as.symbol(name_metadata))
      )
      )
  )
}

tar_decontaminate_phyloseq <- function(phyloseq, neg = NULL, batch = NULL) {
  if(is.null(neg)) { stop("No column defining negative controls was provided!") }
  name_contaminants <- paste0(deparse(quote(phyloseq)), "_contaminants")
  name_decontaminated <- paste0(deparse(quote(phyloseq)), "_decontaminated")
  
  list(
    tar_target_raw(name_contaminants,
                   substitute(
                     decontam::isContaminant(phylo, neg = neg, batch = batch, method = "frequency"),
                     list(phylo = phyloseq, neg = neg, batch = batch)
                   )),
    tar_target_raw(name_decontaminated,
                   substitute(
                     prune_taxa(!contaminants$contaminant, phylo),
                     list(phylo = phyloseq, contaminants = as.symbol(name_contaminants))
                   ))
  )
}
```

_targets.R file:

```r
library(targets)
tar_option_set(packages = c("tidyverse", "phyloseq"))

list(
  ## Read in data
  tar_read_ampliseq(rrna_16s, 
               here::here("path/to/file"),
               here::here("path/to/file"),
               here::here("path/to/file")
               ),
  tar_decontaminate_phyloseq(rrna_16s_phyloseq, "Control")
)
```

```
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: object 'rrna_16s_phyloseq' not found
```

If anyone has any insight, it would be much appreciated - I can't see for the life of me what I've missed!

**Comment by wlandau** [comment]
Looks like one issue is how `tar_decontaminate_phyloseq()` captures the target name. I can reproduce it outside the pipeline:

```r
tar_decontaminate_phyloseq(rrna_16s_phyloseq, "Control")
#> Error: object 'rrna_16s_phyloseq' not found
```

If you want to pass the name as a symbol, `deparse(substitute())` will quote the argument and turn it into a string.

```r
tar_decontaminate_phyloseq <- function(phyloseq, neg = NULL, batch = NULL) {
  target_name <- deparse(substitute(phyloseq))
  ...
}
```

Once you fix this, I recommend checking `tar_visnetwork()` to see if the dependency graph is correct.

Unfortunately I cannot test the pipeline because I do not have the `phyloseq` package, and I do not have the physical files it seems to require. If you have more trouble, it would help if the example code were simpler and runnable, as in a reprex.

**Reply by prototaxites** [reply]
Apologies for the late reply to this, I had to put this down for a couple of weeks. Thanks for the tips, and apologies for the non-reproducable code - I'll make sure to build a better example next time :) 

I managed to fix this today - I switched out the code as you suggested above, which didn't fix things. What I had to do was change `substitute(..., list(phylo = phyloseq))` to `substitute(..., list(phylo = as.symbol(phyloseq)))`, and now the function works. 

I definitely need to brush up on my R metaprogramming skills before I can progress much further with this - can you recommend any good resources for learning the basics?

**Reply by wlandau** [reply]
For meta programming, I have some pointers at https://wlandau.github.io/targetopia/contributing.html. https://adv-r.hadley.nz/expressions.html is more comprehensive.
